pkg/updatestrategy/drain.go

```diff
@@ -0,0 +1,450 @@
+		if err != nil {
+			return false, err
+		}
+
+		// try evicting normally
+		err = evictPod(m.kube, m.logger, pod)
+		if err == nil {
+			return false, nil
+		}
+		if !isPDBViolation(err) {
+			return false, err
+		}
+
+		// PDB violation, check if we can force terminate the pod
+		m.pdbViolated(pod)
+
+		forceTerminate, err := m.forceTerminationAllowed(pod, time.Now(), drainStart, lastForcedTermination)
+		if forceTerminate {
+			err = deletePod(m.kube, m.podLogger(pod), pod)
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (m *KubernetesNodePoolManager) forceTerminationAllowed(pod v1.Pod, now, drainStart, lastForcedTermination time.Time) (bool, error) {
+	// too early to start force terminating
+	if now.Before(drainStart.Add(m.drainConfig.ForceEvictionGracePeriod)) {
+		m.podLogger(pod).Debugf("Won't force terminate (node in grace period)")
+		return false, nil
+	}
+
+	// we've recently force killed a pod
+	if now.Before(lastForcedTermination.Add(m.drainConfig.ForceEvictionInterval)) {
+		m.podLogger(pod).Debugf("Won't force terminate (recently force terminated a pod)")
+		return false, nil
+	}
+
+	// pod too young
+	if now.Before(pod.GetCreationTimestamp().Add(m.drainConfig.MinPodLifetime)) {
+		m.podLogger(pod).Debugf("Won't force terminate (pod too young)")
+		return false, nil
+	}
+
+	// find all other pods matched by the same PDBs
+	allPods, err := m.getPodsByNamespace(pod.GetNamespace())
+	if err != nil {
+		return false, err
+	}
+	allPdbs, err := m.getPDBsbyNamespace(pod.GetNamespace())
+	if err != nil {
+		return false, err
+	}
+	siblingPods := findSiblingPods(m.logger, pod, allPods.Items, allPdbs.Items)
+
+	// check if PDB siblings are old enough
+	siblingsOldEnough := true
+	for _, siblingPod := range siblingPods {
+		waitTime := m.drainConfig.MinUnhealthyPDBSiblingLifetime
+		if podReady(siblingPod) {
+			waitTime = m.drainConfig.MinHealthyPDBSiblingLifetime
+		}
+		if now.Before(siblingPod.GetCreationTimestamp().Add(waitTime)) {
+			siblingsOldEnough = false
+			break
+		}
+	}
+	if !siblingsOldEnough {
+		var siblingNames []string
+		for _, siblingPod := range siblingPods {
+			siblingNames = append(siblingNames, siblingPod.GetName())
+		}
+		m.podLogger(pod).Debugf("Won't force terminate (siblings %s too young)", strings.Join(siblingNames, ", "))
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func podReady(pod v1.Pod) bool {
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == v1.PodReady {
+			return condition.Status == v1.ConditionTrue
+		}
+	}
+	return false
+}
+
+func findSiblingPods(logger *log.Entry, pod v1.Pod, allPods []v1.Pod, allPdbs []policy.PodDisruptionBudget) []v1.Pod {
+	var siblingSelectors []labels.Selector
+	for _, pdb := range allPdbs {
+		selector, err := metav1.LabelSelectorAsSelector(pdb.Spec.Selector)
+		if err != nil {
+			logger.Debugf("pdb %s/%s has an invalid selector: %v", pod.GetNamespace(), pdb.GetName(), err)
```

Comment:
Why not use `m.podLogger(pod).Debugf` ?

Reply:
It's a function, not a method, which makes it a bit easier to test. The logger passed to it already is a `podLogger` which achieves the same behaviour in the end.